gloo/allreduce.cc

```diff
@@ -271,7 +279,7 @@ void ring(
     return result;
   };
 
-  for (auto i = 0; i < numSegments; i++) {
+  for (auto i = 0; i < (numSegments - numSegmentsPerRank + 2); i++) {
```

Comment:
Shall we have some comments to explain this? I don't fully understand why we have `- numSegmentsPerRank + 2` here. Where does the 2 come from? I was expecting the two if clauses be 1) `if (i >= numSegmentsPerRank)` and 2) `if (i < (numSegments - numSegmentsPerRank))`, meaning that we issue send/recv ops in the first `numSegments - numSegmentsPerRank` iterations, and do reduce in the last `numSegments - numSegmentsPerRank` iterations. Did I miss anything here?

Reply:
The `2` is due to pipelining. We have 2 segments in flight and then wait on the first one to complete, compute a reduction, and put the next one in flight.

I added comments to explain the rationale.

Btw, there is a way to rewrite the loop for this constant to be `1`, but it doesn't simplify anything IMO, so I rather leave it as is, with the additional explanation.